app/models/billing/page_unlock_promo_code.rb

```diff
@@ -0,0 +1,32 @@
+class PageUnlockPromoCode < ApplicationRecord
+  has_many :promotions
+  has_many :users, -> { distinct }, through: :promotions
+
+  def unlocked_page_types
+    page_types.split('|') & Rails.application.config.content_types[:all].map(&:name)
+  end
+
+
+  def activate!(user)
+    return false unless uses_remaining > 0
+
+    # Make sure this user hasn't activated this promo code before
+    return false unless user.present?
+    return false if users.pluck(:id).include?(user.id)
+
+    # Activate!
+    # technically two requests at the same time could still double-activate at 
+    # 1 use left, but this seems not worth it to fix
+    update(uses_remaining: uses_remaining - 1)
```

Comment:
How many activates can a use have? Why not if it’s some set number why not pre-populate the table? So that one row is one activation. Then you can just update rows that have not been flipped to activated. This could remove this race condition as you’ll get transaction isolation.

Reply:
This is an interesting idea (and might make more sense later), but I think collapsing a bunch of promo codes into a single row makes it easier to reason with in a few key situations (especially at the start/release):

* Seeing how many people have activated a particular code
* Seeing which codes are being used the most
* Adding more uses to specific codes
* Disabling specific codes

I do miss out on per-activation metrics compared to having N rows e.g. adding `activated_by`, `activated_at`, etc to see who/when codes are activating, but I think that'd be easy to transition to later (by just starting to populate N rows with `uses_remaining` = 1). 

Obviously, I could technically do all of the above with either approach, but I think this one is nice and simple to start off with to make sure everything works. :)

Reply:
Simple is nice 👍